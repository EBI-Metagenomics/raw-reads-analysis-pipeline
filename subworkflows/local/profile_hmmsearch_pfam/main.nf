include { SEQKIT_TRANSLATE } from '../../../modules/nf-core/seqkit/translate/main'
include { FASTAEMBEDLENGTH } from '../../../modules/local/fastaembedlength/main'
include { HMMER_HMMSEARCH } from '../../../modules/nf-core/hmmer/hmmsearch/main'
include { PARSEHMMSEARCHCOVERAGE } from '../../../modules/local/parsehmmsearchcoverage/main'
include { COMBINEHMMSEARCHTBL } from '../../../modules/local/combinehmmsearchtbl/main'

workflow PROFILE_HMMSEARCH_PFAM {

    take:
    reads_fasta
    pfam_db

    main:
    ch_versions = Channel.empty()
    FASTAEMBEDLENGTH(reads_fasta, file("${projectDir}/bin/fastx_embed_length.py"))
    SEQKIT_TRANSLATE(FASTAEMBEDLENGTH.out.fasta)

    ch_chunked_pfam_in = SEQKIT_TRANSLATE.out.fastx
        .flatMap{ meta, fasta ->
             def chunks = fasta.splitFasta(file: true, size: params.hmmsearch_chunksize)
             chunks.collect{ chunk -> tuple(groupKey(meta, chunks.size()), chunk) }
         }
        .combine(pfam_db)
        .map{ meta, reads, db -> [meta, db, reads, true, true, true] }

    HMMER_HMMSEARCH(ch_chunked_pfam_in)
    ch_versions = ch_versions.mix(HMMER_HMMSEARCH.out.versions)

    COMBINEHMMSEARCHTBL(
        HMMER_HMMSEARCH.out.domain_summary.groupTuple()
    )

    PARSEHMMSEARCHCOVERAGE(COMBINEHMMSEARCHTBL.out.concatenated_result, file("${projectDir}/bin/hmmer_domtbl_parse_coverage.py"))
    ch_versions = ch_versions.mix(PARSEHMMSEARCHCOVERAGE.out.versions)

    emit:
    profile  = PARSEHMMSEARCHCOVERAGE.out.tsv
    versions = ch_versions                     // channel: [ versions.yml ]
}

