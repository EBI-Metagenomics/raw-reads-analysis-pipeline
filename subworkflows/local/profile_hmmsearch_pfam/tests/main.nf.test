nextflow_workflow {

    name "Test Subworkflow PROFILE_HMMSEARCH_PFAM"
    script "../main.nf"
    workflow "PROFILE_HMMSEARCH_PFAM"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/profile_hmmsearch_pfam"


    test("mini sample 1 with some matches should match snapshot") {

        when {
            workflow {
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test1', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/profile_hmmsearch_pfam/ERR10889147.mini.fasta.gz", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/pfam_mini/Pfam-A.hmm", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def profile = workflow.out.profile.collect{ meta, fp -> [meta, file(fp).text.readLines().sort()] }

            assert workflow.success
            assert snapshot(
                profile,
		workflow.out.versions
	    ).match()
        }
    }
    
    test("mini sample 2 with some matches should match snapshot") {

        when {
            workflow {
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test1', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/profile_hmmsearch_pfam/ERR10889056.mini.fasta.gz", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/pfam_mini/Pfam-A.hmm", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def profile = workflow.out.profile.collect{ meta, fp -> [meta, file(fp).text.readLines().sort()] }

            assert workflow.success
            assert snapshot(
                profile,
		workflow.out.versions
	    ).match()
        }
    }
    
    test("mini sample 1 with no matches should match snapshot") {

        when {
            workflow {
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test1', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/profile_hmmsearch_pfam/ERR10889147.not_matches.fasta.gz", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/pfam_mini/Pfam-A.hmm", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def profile = workflow.out.profile.collect{ meta, fp -> [meta, file(fp).text.readLines().size()] }

            assert workflow.success
            assert snapshot(
                profile,
		workflow.out.versions
	    ).match()
        }
    }
    
    test("mini sample 2 with no matches should match snapshot") {

        when {
            workflow {
                """
                input[0] = Channel.of(
		    [
		        ['id': 'test1', 'single_end': false, 'platform': 'wgs'],
	                file("${projectDir}/tests/fixtures/profile_hmmsearch_pfam/ERR10889056.not_matches.fasta.gz", checkIfExists: true),
		    ]
		)
		input[1] = Channel.value(
		    file("${projectDir}/tests/fixtures/dbs/pfam_mini/Pfam-A.hmm", checkIfExists: true)
		)
                """
            }
        }

        then {
	    def profile = workflow.out.profile.collect{ meta, fp -> [meta, file(fp).text.readLines().size()] }

            assert workflow.success
            assert snapshot(
                profile,
		workflow.out.versions
	    ).match()
        }
    }
    
}
