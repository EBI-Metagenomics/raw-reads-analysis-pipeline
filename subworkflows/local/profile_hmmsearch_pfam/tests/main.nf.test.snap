{
    "mini sample 2 with some matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test1",
                            "single_end": false,
                            "platform": "wgs"
                        }
                    },
                    [
                        "# Query Accession\tRead Count\tCoverage Depth\tCoverage Breadth\tExpected Coverage Breadth\tObserved:Expected Coverage Breadth Ratio",
                        "PF00004.35\t6\t1.549618320610687\t0.7175572519083969\t0.6021338254758717\t1.1916906533880656",
                        "PF00198.28\t1\t0.16810344827586207\t0.16810344827586207\t0.1121773780091837\t1.4985503428516562",
                        "PF00389.36\t3\t0.8731343283582089\t0.5223880597014925\t0.432835939587608\t1.2068962207694833",
                        "PF00393.24\t8\t1.0344827586206897\t0.6275862068965518\t0.4821799058272518\t1.3015602668465285",
                        "PF02096.25\t1\t0.2331288343558282\t0.2331288343558282\t0.1510645996013471\t1.5432393490668563",
                        "PF02737.24\t1\t0.18333333333333332\t0.18333333333333332\t0.12149689094119565\t1.508954936320687",
                        "PF02739.22\t2\t0.6585365853658537\t0.42073170731707316\t0.35561483971802343\t1.1831106588540643",
                        "PF02826.25\t6\t1.3258426966292134\t0.8258426966292135\t0.5560654529582161\t1.4851537570546915",
                        "PF03061.28\t1\t0.4050632911392405\t0.4050632911392405\t0.24333012715521096\t1.6646655959739263",
                        "PF04463.17\t1\t0.2765957446808511\t0.2765957446808511\t0.1757842836214759\t1.5734953033483752",
                        "PF06439.18\t2\t0.39037433155080214\t0.37433155080213903\t0.23600245339422588\t1.5861341499566697",
                        "PF06725.17\t1\t0.4444444444444444\t0.4444444444444444\t0.2624996803309244\t1.6931237549857143",
                        "PF06750.18\t1\t0.42857142857142855\t0.42857142857142855\t0.254855513859219\t1.6816250984005368",
                        "PF07724.20\t6\t1.2951807228915662\t0.5662650602409639\t0.5490914029637832\t1.0312765000225534",
                        "PF07726.17\t2\t0.549618320610687\t0.549618320610687\t0.3104731538596649\t1.7702603712368532",
                        "PF09820.15\t7\t0.8960573476702509\t0.5806451612903226\t0.4402666359321351\t1.3188488836111265",
                        "PF10417.14\t1\t0.6829268292682927\t0.6829268292682927\t0.36513957152430687\t1.8703172225824645",
                        "PF13173.12\t6\t1.7286821705426356\t0.7441860465116279\t0.6336607919131878\t1.1744233760538274",
                        "PF13304.12\t2\t0.22697368421052633\t0.1611842105263158\t0.14748323125332718\t1.092898556375232",
                        "PF13476.12\t2\t0.2935323383084577\t0.1791044776119403\t0.18515111198151069\t0.9673421655162718",
                        "PF13614.12\t7\t1.4124293785310735\t0.6610169491525424\t0.574860733674394\t1.1498731961173538",
                        "PF17862.7\t1\t0.6888888888888889\t0.6888888888888889\t0.3674368368037483\t1.8748498242075586"
                    ]
                ]
            ],
            [
                "versions.yml:md5,2f0836d3c70b78f8882ba96e14d952a7",
                "versions.yml:md5,3b930c014552486aedf1f2adf353beac"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-30T09:41:16.315277002"
    },
    "mini sample 1 with no matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test1",
                            "single_end": false,
                            "platform": "wgs"
                        }
                    },
                    1
                ]
            ],
            [
                "versions.yml:md5,2f0836d3c70b78f8882ba96e14d952a7",
                "versions.yml:md5,3b930c014552486aedf1f2adf353beac"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-30T09:49:53.432687571"
    },
    "mini sample 2 with no matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test1",
                            "single_end": false,
                            "platform": "wgs"
                        }
                    },
                    1
                ]
            ],
            [
                "versions.yml:md5,2f0836d3c70b78f8882ba96e14d952a7",
                "versions.yml:md5,3b930c014552486aedf1f2adf353beac"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-30T09:51:41.29316408"
    },
    "mini sample 1 with some matches should match snapshot": {
        "content": [
            [
                [
                    {
                        "groupSize": 1,
                        "groupTarget": {
                            "id": "test1",
                            "single_end": false,
                            "platform": "wgs"
                        }
                    },
                    [
                        "# Query Accession\tRead Count\tCoverage Depth\tCoverage Breadth\tExpected Coverage Breadth\tObserved:Expected Coverage Breadth Ratio",
                        "PF00004.35\t5\t0.9770992366412213\t0.5877862595419847\t0.4654000080605616\t1.2629700244128428",
                        "PF00389.36\t2\t0.582089552238806\t0.35074626865671643\t0.3243920843754171\t1.0812417612841618",
                        "PF00393.24\t5\t0.5620689655172414\t0.41724137931034483\t0.3158579897336943\t1.3209777585874234",
                        "PF01367.25\t1\t0.41237113402061853\t0.41237113402061853\t0.24693936662683835\t1.6699286940497093",
                        "PF02096.25\t1\t0.2392638036809816\t0.2392638036809816\t0.15461399517800434\t1.5474912436323855",
                        "PF02826.25\t5\t1.101123595505618\t0.5674157303370787\t0.5006786127368137\t1.1332933261028786",
                        "PF02872.23\t3\t0.6075949367088608\t0.43670886075949367\t0.33504572556865386\t1.3034306288142994",
                        "PF03061.28\t1\t0.45569620253164556\t0.45569620253164556\t0.26785212691174554\t1.7012976816189056",
                        "PF07724.20\t6\t1.3433734939759037\t0.6566265060240963\t0.5599763581259221\t1.1725968364479427",
                        "PF07726.17\t4\t1.1221374045801527\t0.6564885496183206\t0.5063018678576134\t1.2966346586794422",
                        "PF09821.14\t1\t0.31932773109243695\t0.31932773109243695\t0.19913997334881972\t1.6035340656247485",
                        "PF12846.13\t2\t0.22099447513812154\t0.22099447513812154\t0.14398464627951812\t1.534847505261803",
                        "PF13173.12\t4\t1.054263565891473\t0.7286821705426356\t0.48777884238762426\t1.4938781825300496",
                        "PF13245.12\t1\t0.2835820895522388\t0.2835820895522388\t0.17966580348201655\t1.5783865602483649",
                        "PF13304.12\t3\t0.34539473684210525\t0.25\t0.21294379053462142\t1.1740187369274513",
                        "PF13476.12\t4\t0.6218905472636815\t0.22885572139303484\t0.34091212102601176\t0.6713041493047208",
                        "PF13555.12\t3\t1.1639344262295082\t0.5573770491803278\t0.5171999349243717\t1.0776819785598604",
                        "PF13604.12\t1\t0.19895287958115182\t0.19895287958115182\t0.13091887127109814\t1.519665405372869",
                        "PF13614.12\t7\t1.3954802259887005\t0.6610169491525424\t0.5712827592687935\t1.157074913303183",
                        "PF16193.11\t1\t0.43209876543209874\t0.43209876543209874\t0.25656373903783714\t1.6841770666913078",
                        "PF17862.7\t1\t0.8222222222222222\t0.8222222222222222\t0.4157983136284613\t1.9774544419074365",
                        "PF20789.3\t1\t0.2857142857142857\t0.2857142857142857\t0.1808454587914402\t1.579880897334477"
                    ]
                ]
            ],
            [
                "versions.yml:md5,2f0836d3c70b78f8882ba96e14d952a7",
                "versions.yml:md5,3b930c014552486aedf1f2adf353beac"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-05-30T09:34:16.809032813"
    }
}